test/integration/job/job_test.go

```diff
@@ -1820,6 +1821,127 @@ func TestJobPodReplacementPolicy(t *testing.T) {
+			},
+			wantTerminating: ptr.To(podCount),
+			wantActive:      int(podCount),
+			wantFailed:      int(podCount),
+		},
+	}
+	for name, tc := range cases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobPodReplacementPolicy, true)()
+
+			closeFn, restConfig, clientSet, ns := setup(t, "pod-replacement-policy")
+			defer closeFn()
+			ctx, cancel := startJobControllerAndWaitForCaches(restConfig)
+			defer func() {
+				cancel()
+			}()
+			resetMetrics()
+
+			jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &batchv1.Job{
+				Spec: *tc.jobSpec,
+			})
+			if err != nil {
+				t.Fatalf("Failed to create Job: %v", err)
+			}
+			jobClient := clientSet.BatchV1().Jobs(jobObj.Namespace)
+
+			// Wait for pods to start up.
+			err = wait.PollUntilContextTimeout(ctx, 5*time.Millisecond, wait.ForeverTestTimeout, true, func(context.Context) (done bool, err error) {
+				job, err := jobClient.Get(ctx, jobObj.Name, metav1.GetOptions{})
+				if err != nil {
+					return false, err
+				}
+				if job.Status.Active == int32(podCount) {
+					return true, nil
+				}
+				return false, nil
+			})
+			if err != nil {
+				t.Fatalf("Error waiting for Job pods to become active: %v", err)
+			}
+			pods, errList := clientSet.CoreV1().Pods(ns.Namespace).List(ctx, metav1.ListOptions{})
+			if errList != nil {
+				t.Fatalf("Failed to list pods: %v", errList)
+			}
+			updatePod(t, clientSet, pods.Items, func(pod *v1.Pod) {
+				pod.Finalizers = append(pod.Finalizers, "fake.example.com/blockDeletion")
+			})
+			err = clientSet.CoreV1().Pods(ns.Name).DeleteCollection(ctx,
+				metav1.DeleteOptions{},
+				metav1.ListOptions{
+					Limit: 1000,
+				})
+			if err != nil {
+				t.Fatalf("Failed to cleanup Pods: %v", err)
+			}
+
+			podsDelete, errList2 := clientSet.CoreV1().Pods(ns.Namespace).List(ctx, metav1.ListOptions{})
+			if errList != nil {
+				t.Fatalf("Failed to list pods: %v", errList2)
+			}
+			for _, val := range podsDelete.Items {
+				if val.DeletionTimestamp == nil {
+					t.Fatalf("Deletion not registered.")
+				}
+			}
+
+			validateJobsPodsStatusOnly(ctx, t, clientSet, jobObj, podsByStatus{
+				Terminating: tc.wantTerminating,
+				Failed:      tc.wantFailed,
+				Ready:       ptr.To[int32](0),
+			})
+			// Disable the controller and turn off feature.
+			cancel()
+			defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobPodReplacementPolicy, false)()
+			ctx, cancel = startJobControllerAndWaitForCaches(restConfig)
+			validateJobsPodsStatusOnly(ctx, t, clientSet, jobObj, podsByStatus{
+				Active: int(podCount),
+				Failed: int(podCount),
+				Ready:  ptr.To[int32](0),
+			})
+			// Disable the controller and turn feature on again.
+			cancel()
+			defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobPodReplacementPolicy, true)()
```

Comment:
@mimowo this is not quite working right.  

It looks like something is up with the last feature gate.

Reply:
Any ideas?  I don't see a lot of use cases of doing enable -> disable -> enable in the integration tests.

Reply:
No ideas for now. This is an example test which reenables the feature gate: https://github.com/kubernetes/kubernetes/blob/e8abe1af8dcb36f65ef7aa7135d4664b3db90e89/test/integration/job/job_test.go#L763

Reply:
Hmm. Alright I’ll see if I can find out why.

Reply:
Still not quite figuring this out.

So I can get the case of enable -> disable working correctly but when I try to restart the pod and start the new controller, I get bad results.  

I pushed up a change where I create a new job object on the last step and that seems to work fine.  

The main issues I see are that deletion of the pods doesn't seem to work on restarting controllers.  I notice that I lose the deletiontimestamp on restart.  So I tried to send a new delete for the pods and remove the finaiziers but that didn't work.  

WDYT about this?  I don't quite understand what we want to test with this case.

I could also have a rollout and a rollback test rather than a enable -> disable -> enable.  This seems to have some issues.

Reply:
I can take a deeper look, but for now I'm wondering if (and why) you want to have pods with deletiontimestamp during restart, this sounds tricky, IMO. 

I think we could have a test that shows how recreation is handled when enabled, disabled, and enabled again. And there are no pods (or at least without deletion timestamps) when transitioning between enabled, disabled and enabled again.

Reply:
If there are no deletion timestamps, then that means on the enabled there would be no terminating pods.  

I can do that but it seems to not be what we want to test with this.  Counting the terminating pods is crucial to this feature as that is what we use to flag a mismatch.

Reply:
I did but I realize that maybe I didn't do it right.  

I don't know how to change the feature gate without recreating the cluster.  

So my kind test was create kind cluster with feature on, verfy, delete kind cluster, and then create new cluster with it off.  I don't think that is correct actually.  

How do you change the feature gate of the cluster in kind?  I tried to edit the control plane pod but I am unable to edit the pod.

Reply:
> How do you change the feature gate of the cluster in kind? I tried to edit the control plane pod but I am unable to edit the pod.

I usually edit control plane pod manifests, for example:
```sh
docker exec -it per-index-control-plane sed -i 's/JobBackoffLimitPerIndex=true/JobBackoffLimitPerIndex=false/' /etc/kubernetes/manifests/kube-controller-manager.yaml
docker exec -it per-index-control-plane sed -i 's/JobBackoffLimitPerIndex=true/JobBackoffLimitPerIndex=false/' /etc/kubernetes/manifests/kube-apiserver.yaml
```
Changes in these files are picked automatically and the pods are restarted.

Reply:
I checked the test locally in vs code and added the following to list the pods (at the end of the test):
```
for _, val := range podsDelete.Items {
  t.Logf("POD: %v, hasDeletionTimestamp=%v, creationTimestamp=%v, finalizers=%v", val.Name, val.DeletionTimestamp != nil, val.CreationTimestamp, val.Finalizers)
}
```
I'm getting logged:
```
...integration/job/job_test.go:1902: POD: test-job-2bxzz, hasDeletionTimestamp=false, creationTimestamp=2023-10-04 09:20:49 +0200 CEST, finalizers=[batch.kubernetes.io/job-tracking]
...integration/job/job_test.go:1902: POD: test-job-b4gfk, hasDeletionTimestamp=true, creationTimestamp=2023-10-04 09:20:29 +0200 CEST, finalizers=[fake.example.com/blockDeletion]
...integration/job/job_test.go:1902: POD: test-job-blk46, hasDeletionTimestamp=true, creationTimestamp=2023-10-04 09:20:29 +0200 CEST, finalizers=[fake.example.com/blockDeletion]
...integration/job/job_test.go:1902: POD: test-job-l6rt6, hasDeletionTimestamp=false, creationTimestamp=2023-10-04 09:20:49 +0200 CEST, finalizers=[batch.kubernetes.io/job-tracking]
...integration/job/job_test.go:1906: Deletion not registered.
```
So, I think you just get the new pods created (the creation timestamps clearly split in two batches, and the new batch of pods does not have the finalizer. 

This seems expected given you don't specify the `podReplacementPolicy` (at least in this version of the test). I'm not sure what is the testing scenario.

Reply:
For the last case, I just want to verify that the deletion timestamp was not there.  I didnt realize that I was checking more pods than 2.

Reply:
figured out all these issues. resolving.